hbase-server/src/test/java/org/apache/hadoop/hbase/io/hfile/TestHFileWriterWithRowIndexV1DataEncoder.java

```diff
@@ -0,0 +1,103 @@
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.CellComparatorImpl;
+import org.apache.hadoop.hbase.HBaseClassTestRule;
+import org.apache.hadoop.hbase.HBaseTestingUtility;
+import org.apache.hadoop.hbase.KeyValue;
+import org.apache.hadoop.hbase.io.encoding.DataBlockEncoding;
+import org.apache.hadoop.hbase.testclassification.IOTests;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Category({ IOTests.class, MediumTests.class })
+public class TestHFileWriterWithRowIndexV1DataEncoder {
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestHFileWriterWithRowIndexV1DataEncoder.class);
+
+  private static final Logger LOG =
+    LoggerFactory.getLogger(TestHFileWriterWithRowIndexV1DataEncoder.class);
+
+  private static final HBaseTestingUtility TEST_UTIL = new HBaseTestingUtility();
+
+  private Configuration conf;
+  private FileSystem fs;
+  private DataBlockEncoding dataBlockEncoding;
+
+  @Before
+  public void setUp() throws IOException {
+    conf = TEST_UTIL.getConfiguration();
+    fs = FileSystem.get(conf);
+    dataBlockEncoding = DataBlockEncoding.ROW_INDEX_V1;
+  }
+
+  @Test
+  public void testBlockCountWritten() throws IOException {
+    Path hfilePath = new Path(TEST_UTIL.getDataTestDir(), "testHFileFormatV3");
+    final int entryCount = 10000;
+    writeDataAndReadFromHFile(hfilePath, entryCount);
+  }
+
+  private void writeDataAndReadFromHFile(Path hfilePath, int entryCount) throws IOException {
+    HFileContext context =
+      new HFileContextBuilder().withBlockSize(1024).withDataBlockEncoding(dataBlockEncoding)
+        .withCellComparator(CellComparatorImpl.COMPARATOR).build();
+    CacheConfig cacheConfig = new CacheConfig(conf);
+    HFile.Writer writer =
+      new HFile.WriterFactory(conf, cacheConfig).withPath(fs, hfilePath).withFileContext(context)
+        .create();
+
+    List<KeyValue> keyValues = new ArrayList<>(entryCount);
+
+    writeKeyValues(entryCount, writer, keyValues);
+
+    FSDataInputStream fsdis = fs.open(hfilePath);
+
+    long fileSize = fs.getFileStatus(hfilePath).getLen();
+    FixedFileTrailer trailer = FixedFileTrailer.readFromStream(fsdis, fileSize);
+
+    // HBASE-23788
+    // kv size = 24 bytes
+    // per row encoded data written = (4 (Row index) + 24 (Cell size) + 1 (MVCC)) bytes = 29 bytes
+    // creating block size of (29 * 36) bytes = 1044 bytes
+    // Number of blocks = ceil((29 * 10000) / 1044) = 278
+    // Without the patch it would have produced 244 blocks
+    Assert.assertEquals(278, trailer.getDataIndexCount());
```

Comment:
LGTM.  But I think you can also check the block size and in the comment just say what was the block count previously .

Reply:
I have mentioned the previous block count in line 79. I will check if I assert the size of the block.